File: src/sources/scale-exporter/reader.ts
import { readdir, readFile } from "node:fs/promises";
import path from "node:path";

import { DateTime } from "luxon";
import { z } from "zod";

import type { ScaleExporterConfig } from "../../config/index.js";
import type {
  MeasurementKind,
  MeasurementReading,
  MeasurementUnit,
} from "../../domain/index.js";

const exporterKinds = [
  "weight",
  "bodyTemperature",
  "bloodPressureSystolic",
  "bloodPressureDiastolic",
  "heartRate",
] as const;

const exporterKindToDomainKind = {
  weight: "weight",
  bodyTemperature: "body_temperature",
  bloodPressureSystolic: "blood_pressure_systolic",
  bloodPressureDiastolic: "blood_pressure_diastolic",
  heartRate: "pulse",
} as const satisfies Record<(typeof exporterKinds)[number], MeasurementKind>;

const readingLineSchema = z.object({
  measuredAt: z.string().min(1),
  kind: z.enum(exporterKinds),
  value: z.number(),
  unit: z.enum(["kg", "celsius", "mmHg", "bpm"]),
  source: z.string().trim().min(1),
});

const fileNamePattern =
  /^scale_exporter_(\d{4}-\d{2}-\d{2})_(apple-health|google-fit)_(\d{3})\.jsonl$/;
const nearMissFileNamePattern = /^scale_exporter_(\d{4}-\d{2}-\d{2})_.+\.jsonl$/;
const finderCopyPattern = /^(.+\.jsonl)のコピー\d*$/;

export type InputAnomalyReason = "file-name-pattern-mismatch";

export interface InputAnomalyCandidate {
  readonly name: string;
  readonly reason: InputAnomalyReason;
}

export interface ScaleExporterFileClassification {
  readonly targetFileNames: readonly string[];
  readonly inputAnomalyCandidates: readonly InputAnomalyCandidate[];
}

export class ScaleExporterFileError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ScaleExporterFileError";
  }
}

export async function readScaleExporterMeasurements(
  config: ScaleExporterConfig,
  referenceTime: Date,
  timeZone: string,
): Promise<MeasurementReading[]> {
  const targetDate = DateTime.fromJSDate(referenceTime, { zone: timeZone })
    .toFormat("yyyy-MM-dd");

  const fileNames = await listTargetFiles(config.outputDir, targetDate);
  const readings: MeasurementReading[] = [];
  const seen = new Set<string>();

  for (const fileName of fileNames) {
    const filePath = path.join(config.outputDir, fileName);
    const text = await readFile(filePath, "utf8");
    const lines = text.split("\n");

    for (const [index, line] of lines.entries()) {
      if (line.trim().length === 0) {
        continue;
      }

      const reading = parseScaleExporterReadingLine(line, fileName, index + 1);
      const dedupKey = [
        reading.measuredAt,
        reading.kind,
        reading.value,
        reading.source,
      ].join("|");
      if (seen.has(dedupKey)) {
        continue;
      }

      seen.add(dedupKey);
      readings.push(reading);
    }
  }

  return readings;
}

async function listTargetFiles(
  outputDir: string,
  targetDate: string,
): Promise<readonly string[]> {
  let entries: string[];
  try {
    entries = await readdir(outputDir);
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code === "ENOENT") {
      return [];
    }
    throw error;
  }

  return classifyScaleExporterFileNames(entries, targetDate).targetFileNames;
}

export function isScaleExporterTargetFile(name: string, targetDate: string): boolean {
  const match = fileNamePattern.exec(name.normalize("NFC"));
  return match !== null && match[1] === targetDate;
}

export function classifyScaleExporterFileNames(
  names: readonly string[],
  targetDate: string,
): ScaleExporterFileClassification {
  const targetFileNames: string[] = [];
  const inputAnomalyCandidates: InputAnomalyCandidate[] = [];
  const anomalyCandidateKeys = new Set<string>();
  for (const name of names) {
    const comparisonName = name.normalize("NFC");
    if (isScaleExporterTargetFile(comparisonName, targetDate)) {
      targetFileNames.push(name);
      continue;
    }
    const copiedName = finderCopyPattern.exec(comparisonName)?.[1];
    if (copiedName !== undefined && isScaleExporterTargetFile(copiedName, targetDate)) {
      continue;
    }
    const nearMiss = nearMissFileNamePattern.exec(comparisonName);
    if (nearMiss?.[1] === targetDate) {
      const reason: InputAnomalyReason = "file-name-pattern-mismatch";
      const key = `${name}\u0000${reason}`;
      if (anomalyCandidateKeys.has(key)) {
        continue;
      }
      anomalyCandidateKeys.add(key);
      inputAnomalyCandidates.push({
        name,
        reason,
      });
    }
  }
  return {
    targetFileNames: targetFileNames.sort(),
    inputAnomalyCandidates: inputAnomalyCandidates.sort((left, right) =>
      left.name.localeCompare(right.name),
    ),
  };
}

export function parseScaleExporterReadingLine(
  line: string,
  fileName: string,
  lineNumber: number,
): MeasurementReading {
  let json: unknown;
  try {
    json = JSON.parse(line);
  } catch {
    throw new ScaleExporterFileError(
      `invalid JSON in ${fileName}:${lineNumber}`,
    );
  }

  const parsed = readingLineSchema.safeParse(json);
  if (!parsed.success) {
    throw new ScaleExporterFileError(
      `invalid reading in ${fileName}:${lineNumber}: ${parsed.error.message}`,
    );
  }

  return {
    kind: exporterKindToDomainKind[parsed.data.kind],
    value: parsed.data.value,
    unit: parsed.data.unit as MeasurementUnit,
    measuredAt: parsed.data.measuredAt,
    source: parsed.data.source,
  };
}

File: internal/sources/scaleexporter/reader.go
package scaleexporter

import (
	"encoding/json"
	"errors"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"

	"github.com/kappaseijin/scale2sheet4go/internal/domain"
)

type Config struct {
	OutputDir string
}

type InputAnomalyReason string

const InputAnomalyFileNamePatternMismatch InputAnomalyReason = "file-name-pattern-mismatch"

type InputAnomalyCandidate struct {
	Name   string
	Reason InputAnomalyReason
}

type FileClassification struct {
	TargetFileNames        []string
	InputAnomalyCandidates []InputAnomalyCandidate
}

type FileError struct {
	Message string
}

func (e *FileError) Error() string { return e.Message }

var (
	fileNamePattern       = regexp.MustCompile(`^scale_exporter_([0-9]{4}-[0-9]{2}-[0-9]{2})_(apple-health|google-fit)_([0-9]{3})\.jsonl$`)
	nearMissFilePattern   = regexp.MustCompile(`^scale_exporter_([0-9]{4}-[0-9]{2}-[0-9]{2})_.+\.jsonl$`)
	finderCopyFilePattern = regexp.MustCompile(`^(.+\.jsonl)のコピー[0-9]*$`)
)

var exporterKindToDomainKind = map[string]domain.MeasurementKind{
	"weight":                 domain.KindWeight,
	"bodyTemperature":        domain.KindBodyTemperature,
	"bloodPressureSystolic":  domain.KindBloodPressureSystolic,
	"bloodPressureDiastolic": domain.KindBloodPressureDiastolic,
	"heartRate":              domain.KindPulse,
}

var validUnits = map[string]domain.MeasurementUnit{
	"kg":      domain.UnitKg,
	"celsius": domain.UnitCelsius,
	"mmHg":    domain.UnitMmHg,
	"bpm":     domain.UnitBPM,
}

func ReadMeasurements(config Config, referenceTime time.Time, timeZone *time.Location) ([]domain.MeasurementReading, error) {
	if timeZone == nil {
		timeZone = time.UTC
	}
	targetDate := referenceTime.In(timeZone).Format("2006-01-02")
	fileNames, err := listTargetFiles(config.OutputDir, targetDate)
	if err != nil {
		return nil, err
	}

	readings := make([]domain.MeasurementReading, 0)
	seen := make(map[string]struct{})
	for _, fileName := range fileNames {
		filePath := filepath.Join(config.OutputDir, fileName)
		content, err := os.ReadFile(filePath)
		if err != nil {
			return nil, err
		}
		for lineNumber, line := range strings.Split(string(content), "\n") {
			if strings.TrimSpace(line) == "" {
				continue
			}
			reading, err := ParseReadingLine(line, fileName, lineNumber+1)
			if err != nil {
				return nil, err
			}
			key := fmt.Sprintf("%s|%s|%.17g|%s", reading.MeasuredAt, reading.Kind, reading.Value, reading.Source)
			if _, exists := seen[key]; exists {
				continue
			}
			seen[key] = struct{}{}
			readings = append(readings, reading)
		}
	}
	return readings, nil
}

func listTargetFiles(outputDir, targetDate string) ([]string, error) {
	entries, err := os.ReadDir(outputDir)
	if errors.Is(err, fs.ErrNotExist) {
		return []string{}, nil
	}
	if err != nil {
		return nil, err
	}
	names := make([]string, 0, len(entries))
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		names = append(names, entry.Name())
	}
	return ClassifyFileNames(names, targetDate).TargetFileNames, nil
}

func IsTargetFile(name, targetDate string) bool {
	match := fileNamePattern.FindStringSubmatch(name)
	return match != nil && match[1] == targetDate
}

func ClassifyFileNames(names []string, targetDate string) FileClassification {
	targetFiles := make([]string, 0)
	anomalies := make([]InputAnomalyCandidate, 0)
	seenAnomalies := make(map[string]struct{})
	for _, name := range names {
		comparisonName := name
		if IsTargetFile(comparisonName, targetDate) {
			targetFiles = append(targetFiles, name)
			continue
		}
		if copied := finderCopyFilePattern.FindStringSubmatch(comparisonName); copied != nil && IsTargetFile(copied[1], targetDate) {
			continue
		}
		nearMiss := nearMissFilePattern.FindStringSubmatch(comparisonName)
		if nearMiss == nil || nearMiss[1] != targetDate {
			continue
		}
		key := name + "\x00" + string(InputAnomalyFileNamePatternMismatch)
		if _, exists := seenAnomalies[key]; exists {
			continue
		}
		seenAnomalies[key] = struct{}{}
		anomalies = append(anomalies, InputAnomalyCandidate{
			Name:   name,
			Reason: InputAnomalyFileNamePatternMismatch,
		})
	}
	sort.Strings(targetFiles)
	sort.Slice(anomalies, func(i, j int) bool { return anomalies[i].Name < anomalies[j].Name })
	return FileClassification{TargetFileNames: targetFiles, InputAnomalyCandidates: anomalies}
}

func ParseReadingLine(line, fileName string, lineNumber int) (domain.MeasurementReading, error) {
	var value struct {
		MeasuredAt string  `json:"measuredAt"`
		Kind       string  `json:"kind"`
		Value      float64 `json:"value"`
		Unit       string  `json:"unit"`
		Source     string  `json:"source"`
	}
	if err := json.Unmarshal([]byte(line), &value); err != nil {
		return domain.MeasurementReading{}, &FileError{Message: fmt.Sprintf("invalid JSON in %s:%d", fileName, lineNumber)}
	}
	kind, kindOK := exporterKindToDomainKind[value.Kind]
	unit, unitOK := validUnits[value.Unit]
	if value.MeasuredAt == "" || !kindOK || !unitOK || strings.TrimSpace(value.Source) == "" {
		return domain.MeasurementReading{}, &FileError{Message: fmt.Sprintf("invalid reading in %s:%d", fileName, lineNumber)}
	}
	return domain.MeasurementReading{
		Kind:       kind,
		Value:      value.Value,
		Unit:       unit,
		MeasuredAt: value.MeasuredAt,
		Source:     value.Source,
	}, nil
}
